### ISSUE TITLE
invoking spconv `SparseConv3d` `SparseConv3d` functions in C++ 

### ISSUE THREAD

**Issue by Jackiq** [traveller59/spconv#13]
Yan,great works! 
I want to use some spconv module in caffe/c++ and I found no ` SparseConv3d` `SubMConv3d` 's declaration in .hpp and I have no idea how to use `SparseConv3d` `SubMConv3d` such functions in c++ code.
Any suggestions would be appreciated.

**Comment by traveller59** [comment]
The sparse convolution is consisted by two step: generate indices (like im2col in conv) and perform gather-gemm-scatter based on indices. the indice generation function and conv function both in include/spconv/spconv_ops.h and are written in torch c++ framework. if you want to use them in caffe (I'm not familiar with it) or other deeplearning frameworks, you need to write your version of getIndicePairs, indiceConv and indiceConvBackward based on my functors.
There are some limits of my code:
1. the batch is represented as batch index in indices. this may be a problem when using in frameworks with static shape such as TensorRT and mxnet. I don't know whether caffe has this limit. you may need to modify my functors to batch-axis format. for example, the indice need to have shape [batchSize, maxNumOfPointPerBatch, ndim] and indicePairs need to have shape [batchSize, kernelVolume, 2, maxNumOfOutPointsPerBatch]. the kernel code in indice.cu and reordering.cu also need to be modified. The chosen of maxNumOfPointPerBatch and maxNumOfOutPointsPerBatch is also a problem.
2. The "to dense" operation is implemented in pure-pytorch code. you need to write a C++ version of it.
3. The indicePair-reuse mechanism is implemented in pure-python, you need to do it manually in C++.
In summary, it's not easy to implement spconv in caffe. you need to be familiar with sparse convolution algorithm and my code. I will write a batch-axis version of spconv but I am busy before 2019-2-3.

**Comment by Jackiq** [comment]
Thanks for quick reply!
I checked the code again and found it is real a huge work to implement such algorithm in Caffe/C++ framework. But i will have a try.Thanks again!
